crates/ra_ide/src/completion/completion_context.rs

```diff
@@ -183,12 +187,20 @@ impl<'a> CompletionContext<'a> {
         // suggest declaration names, see `CompletionKind::Magic`.
         if let Some(name) = find_node_at_offset::<ast::Name>(&file_with_fake_ident, offset) {
             if let Some(bind_pat) = name.syntax().ancestors().find_map(ast::BindPat::cast) {
-                let parent = bind_pat.syntax().parent();
+                let mut parent = bind_pat.syntax().parent();
                 if parent.clone().and_then(ast::MatchArm::cast).is_some()
-                    || parent.and_then(ast::Condition::cast).is_some()
+                    || parent.clone().and_then(ast::Condition::cast).is_some()
                 {
                     self.is_pat_binding = true;
                 }
+
+                while let Some(_) = parent.clone().and_then(ast::TupleStructPat::cast) {
```

Comment:
I don't think we need this additional check here, *any* bind pat can be a path.

```
mod m {
    pub const X: () = ();
}

fn mani() {
    let m::X = (); // this is an irrefutable
}
```

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=681424a92dae15525a40dc3887e6f3e7


But it might be a good idea to not suggest local variables when completing a pattern.

Reply:
So you're saying don't do any traversal to look for a `MatchArm` either?

Also, where are local variables suggested?

Reply:
Locals are added here: https://github.com/rust-analyzer/rust-analyzer/blob/0714a065d578e8b22b0451bfc64378c875fe858f/crates/ra_ide/src/completion/complete_scope.rs#L10, when `res` happens to be [local](https://github.com/rust-analyzer/rust-analyzer/blob/0714a065d578e8b22b0451bfc64378c875fe858f/crates/ra_hir/src/code_model.rs#L1287).

Reply:
>So you're saying don't do any traversal to look for a MatchArm either?

Rigth, we only should check if this is a bind pat (ideally, also check that it lacks `@` and `ref`).

Reply:
I'm a little confused why the `is_pat_binding` is only set when `parent.clone().and_then(ast::MatchArm::cast).is_some() || parent.clone().and_then(ast::Condition::cast).is_some()` this is true. Shouldn't I just be re-using that variable?

Reply:
> I don't think we need this additional check here, _any_ bind pat can be a path.

I think this strategy results in some erroneous completion items as seen in these tests:

- `record_pattern_enum_variant`
- `completes_enum_variants_and_modules`

Should I be restricting somehow by more than just Local scopes?

Reply:
Yeah, I think `record_pattern_enum_variant` should definitely be fixed. It hits another special case: in `Foo { bar }`, bar is both a reference to local and a name of a field. In this case, we should *only* suggest to complete struct fields. 

Long term, we should also indicate if a field init shorthand actually works, by completing `bar: ` if there isn't a eponymous local, and just `bar` otherwise, but this should be out of scope for this PR. 

For `completes_enum_variants_and_modules` I think we are actually fine! There are a couple of false positves (statics and functions), but it is not so important. But yeah, ideally should also filter out statics, functions (and maybe something else? Haven't scrutinized the full set of possibilities yet)